core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -253,50 +264,27 @@ public void retry(String workflowId) {
             throw new ApplicationException(CONFLICT, "Workflow has not started yet");
         }
 
-        // First get the failed task and the cancelled task
-        Task failedTask = null;
-        List<Task> cancelledTasks = new ArrayList<>();
-        for (Task t : workflow.getTasks()) {
-            if (t.getStatus().equals(FAILED)) {
-                failedTask = t;
-            } else if (t.getStatus().equals(CANCELED)) {
-                cancelledTasks.add(t);
-            }
-        }
+        List<Task> failedTasks = getFailedTasksToRetry(workflow);
+
+        List<Task> cancelledTasks = workflow.getTasks().stream()
+                .filter(x->CANCELED.equals(x.getStatus())).collect(Collectors.toList());
 
-        if (failedTask != null && failedTask.getStatus().isSuccessful()) {
+        if (failedTasks.isEmpty()) {
```

Comment:
Can calling isEmpty on failedTasks can cause NPE?

Reply:
I don't believe failedTasks could be null